### PR TITLE
Add: Allow customizing the entire default name

### PR DESCRIPTION
**Pull request by akirak**
Hi, I wanted to customize the entire default name format, so I have added a new customizable variable.

### CODE REVIEW

**Comment by alphapapa** [comment]
Hi Akira,

Thanks, this looks useful.  Would you mind giving me an example of what format you want to use in your config?  Maybe we should provide a few choices in addition to a custom function.

I'll post a few comments in a review.

File: org-bookmark-heading.el [hunk 1]
```diff
@@ -95,11 +100,9 @@ It should take one argument, the path to the file."
   "Return alist for `bookmark-set' for current `org-mode'
 heading.  Set org-id for heading if necessary."
   (let* ((filename (buffer-file-name (org-base-buffer (current-buffer))))
-         (display-filename (funcall org-bookmark-heading-filename-fn filename))
 	 (heading (unless (org-before-first-heading-p)
                     (org-link-display-format (org-get-heading t t))))
-	 (name (concat display-filename (when heading
-                                          (concat ":" heading))))
+         (name (funcall org-bookmark-heading-name-fn filename heading))
```
Comment (alphapapa) [inline org-bookmark-heading.el:105]: I noticed the indentation appears to change in the diff, which clued me into the fact that the original file had some tabs in it.  I don't know how they got there.  :)  I just pushed a commit that untabifies the whole file.  Would you mind rebasing please?

File: org-bookmark-heading.el [hunk 2]
```diff
@@ -130,6 +133,16 @@ heading.  Set org-id for heading if necessary."
 			    (handler . ,handler)
 			    (front-context-string . ,front-context-string)))))
 
+(defun org-bookmark-heading-default-name (filename heading)
```
Comment (alphapapa) [inline org-bookmark-heading.el]: Nitpick: Let's call it `org-bookmark-heading-default-name-fn`.  I'm not sure, but I think that's more consistent with other instances I've seen of the pattern of having a custom function and a default choice.

File: org-bookmark-heading.el [hunk 3]
```diff
@@ -130,6 +133,16 @@ heading.  Set org-id for heading if necessary."
...
+  "Return the default bookmark name.
```
Comment (alphapapa) [inline org-bookmark-heading.el]: Please use this as the docstring:
```
Return bookmark name for FILENAME and HEADING.
Intended as the default value for `org-bookmark-heading-name-fn'.
```

File: org-bookmark-heading.el [hunk 4]
```diff
@@ -130,6 +133,16 @@ heading.  Set org-id for heading if necessary."
...
+
+FILENAME is the path to the file.
+
+Intended as the default value for `org-bookmark-heading-name-fn'."
+  (let ((display-filename (funcall org-bookmark-heading-filename-fn filename)))
```
Comment (alphapapa) [inline org-bookmark-heading.el]: Since this is so simple, I think we can omit binding `display-filename` and just put the `funcall` in the `concat`.  :)

**Review by alphapapa — CHANGES_REQUESTED** [review]
Thank you!

**Comment by akirak** [comment]
Thank you for the code review. I have resolved the issues in your comments.

> Would you mind giving me an example of what format you want to use in your config? 

I have this experimental configuration now:

```emacs-lisp
(use-package org-bookmark-heading
  :after org
  :custom
  (org-bookmark-heading-filename-fn
   (defun akirak/org-bookmark-heading-filename (path)
     (let* ((path (expand-file-name path))
            (project (project-current))
            (dir (abbreviate-file-name (file-name-directory path)))
            (filename (file-name-nondirectory path))
            (root (car-safe (project-roots project))))
       (if root
           (f-relative path (f-parent root))
         path))))
  (org-bookmark-heading-name-fn
   (defun akirak/org-bookmark-heading (path heading)
     (let ((ancestors (org-get-outline-path)))
       (format "\"%s\" in %s%s"
               (substring-no-properties
                (org-link-display-format heading))
               (akirak/org-bookmark-heading-filename path)
               (if ancestors
                   (substring-no-properties
                    (concat ":" (org-format-outline-path
                                 (mapcar #'org-link-display-format ancestors)
                                 nil nil "/")))
                 ""))))))
```

It consists of the heading of the entry, the file path of the file (which actually starts from the project root), and the outline path to the entry in the file, in that order. An example bookmark name produced by this function is as follows:

> "Emacs packages" in notes/aux.org:Open source projects

I made this because some of my files contains multiple entries on the same subjects in different contexts, e.g. trying out Emacs packages and contributing to them. Apparently, bookmark names should contain outline paths, but since they tend to be long, I want to put them at the end, after the file name. I'm pretty sure I tend to complicate things and mess up, but I think it's useful in seeking out opportunities for pushing the limits.

**Comment by alphapapa** [comment]
Hi Akira,

Sorry for overlooking this PR (I'm going through a list of old ones in chronological order).  Are you still interested in pursuing this change?

Thanks.

**Comment by akirak** [comment]
> Are you still interested in pursuing this change?

Maybe no. I rely more on Org links than on Org bookmarks. This package is still sometimes useful, and it should be a built-in feature, but I don't care much about the default bookmark name.

**Comment by alphapapa** [comment]
Ok, since I'm satisfied with the existing behavior, I'll leave this PR closed.  If you or anyone else is interested in it in the future, feel free to comment and we can reopen it.

**Pull request closed without merging**